CONTRIBUTING.md

```diff
@@ -12,3 +12,16 @@ for general information about the project.
 
 After making any changes to .proto files make sure to generate all
 implementation by running `make gen-all`.
+
+## Profiles Compatibility Requirements
+
+The Otel profile format in
+[profiles/v1experimental](opentelemetry/proto/profiles/v1experimental)
+is derived from
+[pprof](https://github.com/google/pprof/tree/main/proto) and is currently
+in experimental state. We would like to continue maintaining compatibility with pprof.
+All changes to profile proto should meet the following compatibility criteria:
+
+- Every valid pprof profile is also a valid Otel profile AND,
+- Every Otel profile is a valid pprof profile, after any new fields introduced in
```

Comment:
We expand if this is not clear enough.

Reply:
That second requirement is interesting. What does "valid" mean here? That it doesn't cause a decoding error? Or that it allows the pprof CLI tool to render flame graph? Arguably there isn't much value in being "valid" if it doesn't enable interoperability.

Reply:
"doesn't cause a decoding error" is not good enough for me. All we would need to accomplish this is to just assign all new field numbers for Otel pprofextended messages and there would be no decoding errors since Protobuf decoder happily ignores unknown fields. This is obviously of no practical value.

What I am aiming for here is the following. If we extend profiles in Otel by new capabilities then those new capabilities obviously can't be understood by pprof. However _ignoring_ those new additional capabilities should still result in a meaningful profile from pprof perspective.

Reply:
We should discuss the meaning of 'meaningful' in the next SIG meeting. Taking into account that the scope of [pprofextended](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/profiles/v1experimental/pprofextended.proto) is larger than [pprof](https://github.com/google/pprof/blob/main/proto/profile.proto), having every OTel profile be a valid and meaningful pprof profile may not be workable.

Quick examples that come to mind include missing symbols due to delayed symbolization (see the [agenda](https://docs.google.com/document/d/19UqPPPlGE83N37MhS93uRlxsP1_wGxQ33Qv6CDHaEp0/edit) for recent discussions around standardizing a symbol upload protocol) and deprecation / non-usage of fields that pprof expects to be present (e.g. [Sample.location_index](https://github.com/open-telemetry/opentelemetry-proto/blob/a05597bff803d3d9405fcdd1e1fb1f42bed4eb7a/opentelemetry/proto/profiles/v1experimental/pprofextended.proto#L228-L230)).

Reply:
I agree, let's discuss in the SIG.

This criterion came up in a meeting last week (suggested by @rsc). This is essentially a forward compatibility requirement. Due to how unknown fields are by default handled by Protobuf deserializers existing pprof tooling can read pprofextended with no code changes. There is certain value in this, although I don't know what practical use cases would benefit from this.

@aalexand also suggested an alternate criteria: "All Otel pprofextended profiles must be convertible to valid pprof profiles" (the precise meaning of "convertible" TBD). This is a less restrictive criteria that gives more flexibility to Otel profiles, but means if we want existing pprof tools to work with Otel profiles we either need a "converter" tool or need to teach pprof to understand the new Otel profiles directly.

There is probably other options that we can discuss.

Reply:
@aalexand it would be great if you can also join the Profiling SIG on May 30 to discuss the options.

Reply:
Yes, I'll try to join.

The main question is basically whether OTel's pprofextended can become a source of truth for existing pprof clients without requiring them to migrate all at once (which is infeasible). For that to happen there needs to be an arbitrarily long period of time where the pprof's subset of pprofextended functionality is semantically equivalent to what pprof has. Today this is not the case - the most prominent example is the change from ID referencing to index referencing, since it's wire and semantics breaking.

An alternative is to simply consider pprofextended a fork of pprof where some sort of converter would exist from pprofextended to pprof format for pprof to be able to read pprofextended format. The converter does not have to be a part of pprof in this case, it could be anywhere. I suspect that Go runtime would probably keep using the current pprof format as well.

Reply:
pprofextended already uses different names than pprof. Is this considered as conflict?
E.g. pprofextended [`Location.mapping_index`](https://github.com/open-telemetry/opentelemetry-proto/blob/a05597bff803d3d9405fcdd1e1fb1f42bed4eb7a/opentelemetry/proto/profiles/v1experimental/pprofextended.proto#L335) vs pprof [`Location.mapping_id`](https://github.com/google/pprof/blob/9665fa269a300a7e802154f4df8c4e9ee970fc5c/proto/profile.proto#L183).

Reply:
>An alternative is to simply consider pprofextended a fork of pprof where some sort of converter would exist from pprofextended to pprof format for pprof to be able to read pprofextended format. The converter does not have to be a part of pprof in this case, it could be anywhere. I suspect that Go runtime would probably keep using the current pprof format as well.

If we do this I think we loose the benefits of having one standard. In that case we may as well give Otel a complete freedom to designs its own format. I would like to avoid this.

>pprofextended already uses different names than pprof. Is this considered as conflict?
E.g. pprofextended [Location.mapping_index](https://github.com/open-telemetry/opentelemetry-proto/blob/a05597bff803d3d9405fcdd1e1fb1f42bed4eb7a/opentelemetry/proto/profiles/v1experimental/pprofextended.proto#L335) vs pprof [Location.mapping_id](https://github.com/google/pprof/blob/9665fa269a300a7e802154f4df8c4e9ee970fc5c/proto/profile.proto#L183).

We should consider 2 aspects of compatibility: wire compatibility and symbolic compatibility. Renaming fields does not impact wire encoding, so if we aim just for wire compatibility renames are fine. However I would argue that we should probably aim for symbolic compatibility and avoid renames, so that we make pprof tooling evolution to newer pprofextended versions easier.

In this particular case it may be nice to rename from `mapping_id` to `mapping_index` but I think names are similar enough that we can live with the old name too.

Reply:
The most important change in this particular case is not just the name change, but the change in meaning:

mapping_id in pprof means "the positive value of the Mapping.id field of the referenced mapping or zero if this is a nil reference".

mapping_index in pprofextended means "the index of the mapping in the repeated Profile.mappings field" (plus rather undefined currently way to express nil references)

This semantics change is going to be fairly hard to deal with in terms of switching existing pprof clients to pprofextended.